app/Http/Middleware/CheckPassword.php

```diff
@@ -34,13 +33,6 @@ public function handle($request, Closure $next)
             } catch (\Exception $e) {
                 return $next($request);
             }
-
-            // Check for the presence of a bearer token if we are not
-            // already authenticated.
-            $authtoken = new AuthToken();
-            if ($authtoken->getUserIdFromRequest()) {
-                Auth::loginUsingId($authtoken->UserId);
-            }
```

Comment:
Can we put this back in? Unless I'm misunderstanding something, removing this block will remove the ability to use bearer tokens to interact with portions of CDash that require authentication.

Reply:
I suppose that we should define what the intended purpose of an authentication token is.  This code previously allowed anyone with a valid token to authenticate themselves when accessing any endpoint, while my PR changes the behavior to only allow authentication tokens to be used for uploading results as you mentioned.

If we were to allow a token to authenticate a user for any endpoint, it would make project-based scoping tricky to define, and could be challenging to maintain.  Is there a use case for allowing tokens to authenticate users for any endpoint?  It seemed like restricting tokens to be write-only was a desirable feature for the original feature request which prompted this PR.  

A better option here would be to use the token scoping options built into Laravel's Sanctum package (mentioned above), but more of the codebase needs to be switched over to Laravel before we can make use of this effectively.  If we used this package, we could easily specify some tokens as submit-only, while other tokens could be used for an API or similar.

Reply:
> Is there a use case for allowing tokens to authenticate users for any endpoint?

Yes, some users have set up scripts to make authenticated API requests using bearer tokens, so this is functionality we need to maintain.

> It seemed like restricting tokens to be write-only was a desirable feature for the original feature request which prompted this PR.

Yes, I understand the feature request to disable "full access" tokens, but it probably makes sense to implement that in a subsequent PR.

> If we were to allow a token to authenticate a user for any endpoint, it would make project-based scoping tricky to define, and could be challenging to maintain

One thought is to distinguish between "scope" and "project", ie make these separate columns. Then the scope column would mean "full access" vs. "submit only", and we would check this field `CheckPassword.php` to make sure "submit only" tokens can't be used for API access.

Reply:
> I think that seems reasonable. Just to make sure I understand correctly, the functionality would be as follows:
> 
> Two types of tokens:
> 
> 1. "full access" -- authenticates any request for the associated user, essentially an alternative to regular cookie-based sessions
> 2. "submit only" -- can optionally be associated with a specific project, but can only be used to submit results
> 
> This PR already adds the two columns and the changes should be relatively straightforward I think.

Yes, that sounds good to me! If it's feasible, I like the idea of the project-specific, submit-only tokens not being associated with a particular user. ie a project-wide submit token that can be rotated periodically.

Reply:
I project-specific submit-only token which isn't associated with a user sounds like a decent idea (for example, to be used on CI machines), but I worry about token identification when it comes time to revoke them.

Consider the scenario:
Bob is a developer for a project and then switches to a new project.  Bob used a submit-only, project-specific token to submit results to this private dashboard and still has access to the token he used to submit results.  How do the administrators identify which token Bob was using?

Having Bob's tokens associated with his userid is useful here because the administrators can identify which tokens he had access to and delete them.  Having them be associated with a userid is also useful because it means that we can check whether a user has access to a project on the backend, meaning that if a user can no longer see a project, they cannot submit results to the project either.

Since the use case for having project-specific, submit-only tokens not be associated with a user is presumably things like CI machines, we could add the ability for only project administrators (and higher user levels) to create submit-only, project-specific tokens which aren't associated with a specific user.  That way, you only have to rotate all of your tokens in the (presumably rare) occurrence of a project administrator leaving instead of any user leaving.

Reply:
This sounds reasonable, I agree with your assessment that per-user, per-project, submit-only tokens could be useful.

As before, please feel free to defer the "user-independent" project submit token idea until a later PR if that makes your life easier.

Reply:
I just pushed changes which re-add the "full access" token functionality, and allow the user to specify a token as "full access", "submit only", or submit only for a specific project.  See the updated PR description for a screenshot of the UI with the various options.

Given that adding user-independent tokens would probably conflict with my upcoming PR to add a UI for managing tokens, I'll hold off for now and create a separate PR to add that functionality.  From a technical perspective, it should be pretty straightforward.